### PR TITLE
Default timeout for IHost.StopAsync. Create Host with DI.

### PR DESCRIPTION
**Pull request by Tratcher**
#1208 This ports the default shutdown timeout from WebHost to Host. It also introduces HostOptions for configuring the value, and changes Host to be instantiated via DI.

I recommend the https://github.com/aspnet/Hosting/pull/1237/files?w=1 view when reviewing Host.cs

### CODE REVIEW

File: src/Microsoft.Extensions.Hosting/HostBuilder.cs [hunk 1]
```diff
@@ -179,6 +179,7 @@ private void CreateServiceProvider()
             services.AddSingleton(_appConfiguration);
             services.AddSingleton<IApplicationLifetime, ApplicationLifetime>();
             services.AddSingleton<IHostLifetime, ProcessLifetime>();
+            services.AddSingleton<IHost, Host>();
```
Comment (muratg) [inline src/Microsoft.Extensions.Hosting/HostBuilder.cs]: Don't these need to be after AddOptions?

Comment (Tratcher) [reply src/Microsoft.Extensions.Hosting/HostBuilder.cs]: It's not order sensitive.

File: src/Microsoft.Extensions.Hosting/Internal/Host.cs [hunk 2]
```diff
@@ -8,22 +8,31 @@
...
+        private readonly IHostLifetime _hostLifetime;
+        private readonly ApplicationLifetime _applicationLifetime;
+        private readonly HostOptions _options;
         private IEnumerable<IHostedService> _hostedServices;
 
-        internal Host(IServiceProvider services)
+        public Host(IServiceProvider services, IApplicationLifetime applicationLifetime, ILogger<Host> logger,
+            IHostLifetime hostLifetime, IOptions<HostOptions> options)
         {
+            if (options == null)
```
Comment (halter73) [inline src/Microsoft.Extensions.Hosting/Internal/Host.cs]: Nit: Might be worth guarding agains and NRE in `StopAsync` by ensuring `options.Value` isn't null. ConsoleLifetime does `options?.Value ?? throw new ArgumentNullException(nameof(options));` which I think is reasonable enough.

File: test/Microsoft.Extensions.Hosting.Tests/HostTests.cs [hunk 3]
```diff
@@ -505,6 +505,67 @@ public void HostCanBeStarted()
         }
 
         [Fact]
+        public async Task HostStopAsyncUsesDefaultTimeoutIfGivenTokenDoesNotFire()
```
Comment (halter73) [inline test/Microsoft.Extensions.Hosting.Tests/HostTests.cs]: Nit: Shorten the `Task.Delay(TimeSpan.FromSeconds(8))` in `HostStopAsyncCanBeCancelledEarly` to be less than 5 seconds.

**Pull request merged**